kernel/src/table_changes/log_replay.rs

```diff
@@ -52,251 +55,258 @@ pub(crate) fn table_changes_action_iter(
-            };
-            visitor.visit_rows_of(actions.as_ref())?;
+        let mut visitor = ProcessCdfCommitVisitor {
+            add_paths: &mut add_paths,
+            remove_dvs: &mut remove_dvs,
+            has_cdc_action: &mut has_cdc_action,
+            protocol: None,
+            metadata: None,
+        };
 
-            if let Some(protocol) = visitor.protocol {
-                ensure_cdf_read_supported(&protocol)
-                    .map_err(|_| Error::change_data_feed_unsupported(commit_file.version))?;
-            }
-            if let Some((schema, configuration)) = visitor.metadata_info {
-                let schema: StructType = serde_json::from_str(&schema)?;
-                // Currently, schema compatibility is defined as having equal schema types. In the
-                // future, more permisive schema evolution will be supported.
-                // See: https://github.com/delta-io/delta-kernel-rs/issues/523
-                require!(
-                    table_schema.as_ref() == &schema,
-                    Error::change_data_feed_incompatible_schema(table_schema, &schema)
-                );
-                let table_properties = TableProperties::from(configuration);
-                check_cdf_table_properties(&table_properties)
-                    .map_err(|_| Error::change_data_feed_unsupported(commit_file.version))?;
+        visitor.visit_rows_of(actions.as_ref())?;
+        let metadata_changed = visitor.metadata.is_some();
+        match (visitor.protocol, visitor.metadata) {
+            (None, None) => {} // no change
+            (protocol, metadata) => {
+                // at least one of protocol and metadata changed, so update the table configuration
+                *table_configuration = TableConfiguration::try_new(
+                    metadata.unwrap_or_else(|| table_configuration.metadata().clone()),
+                    protocol.unwrap_or_else(|| table_configuration.protocol().clone()),
+                    table_configuration.table_root().clone(),
+                    commit_file.version,
+                )?;
+                if !table_configuration.is_cdf_read_supported() {
+                    return Err(Error::change_data_feed_unsupported(commit_file.version));
+                }
             }
         }
-        // We resolve the remove deletion vector map after visiting the entire commit.
-        if has_cdc_action {
-            remove_dvs.clear();
-        } else {
-            // The only (path, deletion_vector) pairs we must track are ones whose path is the
-            // same as an `add` action.
-            remove_dvs.retain(|rm_path, _| add_paths.contains(rm_path));
+        if metadata_changed {
+            require!(
+                table_schema.as_ref() == table_configuration.schema(),
+                Error::change_data_feed_incompatible_schema(
+                    table_schema,
+                    table_configuration.schema()
+                )
```

Comment:
and will this evolve to use the new schema resolution soon?

Reply:
Yep schema evolution is another PR 👍